ClientCore/Extensions/StringExtensions.cs

```diff
@@ -1,25 +1,80 @@
-                return null;
-
-            int index = text.IndexOf("http://", StringComparison.Ordinal);
-            if (index == -1)
-                index = text.IndexOf("ftp://", StringComparison.Ordinal);
-            if (index == -1)
-                index = text.IndexOf("https://", StringComparison.Ordinal);
-
-            if (index == -1)
-                return null; // No link found
-
-            string link = text.Substring(index);
-            return link.Split(' ')[0]; // Nuke any words coming after the link
-        }
+        if (raw.Contains(ESCAPED_INI_NEWLINE_PATTERN, StringComparison.InvariantCulture))
+            throw new ArgumentException($"The string contains an illegal character sequence! ({ESCAPED_INI_NEWLINE_PATTERN})");
+
+        if (raw.Contains(ESCAPED_SEMICOLON, StringComparison.InvariantCulture))
+            throw new ArgumentException($"The string contains an illegal character sequence! ({ESCAPED_SEMICOLON})");
+
+        return raw
+            .Replace(ProgramConstants.INI_NEWLINE_PATTERN, ESCAPED_INI_NEWLINE_PATTERN, StringComparison.InvariantCulture)
+            .Replace(";", ESCAPED_SEMICOLON, StringComparison.InvariantCulture)
+            .Replace(Environment.NewLine, "\n", StringComparison.InvariantCulture)
+            .Replace("\n", ProgramConstants.INI_NEWLINE_PATTERN, StringComparison.InvariantCulture);
     }
-}
+
+    /// <summary>
+    /// Converts an INI-safe string to a normal string.
+    /// </summary>
+    /// <param name="iniString">Input INI string.</param>
+    /// <returns>Regular string.</returns>
+    public static string FromIniString(this string iniString)
+    {
+        return iniString
+            .Replace(ESCAPED_INI_NEWLINE_PATTERN, ProgramConstants.INI_NEWLINE_PATTERN, StringComparison.InvariantCulture)
+            .Replace(ESCAPED_SEMICOLON, ";", StringComparison.InvariantCulture)
+            .Replace(ProgramConstants.INI_NEWLINE_PATTERN, Environment.NewLine, StringComparison.InvariantCulture);
+    }
+
+    /// <summary>
+    /// Looks up a translated string for the specified key.
+    /// </summary>
+    /// <param name="defaultValue">The default string value as a fallback.</param>
+    /// <param name="key">The unique key name.</param>
+    /// <param name="notify">Whether to add this key and value to the list of missing key-values.</param>
+    /// <returns>The translated string value.</returns>
+    /// <remarks>
+    /// This method is referenced by <c>TranslationNotifierGenerator</c> in order to check if the const
+    /// values that are not initialized on client start automatically are missing (via notification
+    /// mechanism implemented down the call chain). Do not change the signature or move the method out
+    /// of the namespace it's currently defined in. If you do - you have to also edit the generator
+    /// source code to match.
+    /// </remarks>
+    public static string L10N(this string defaultValue, string key, bool notify = true)
```

Comment:
I just realized that we need to avoid optional parameters in public method of a library.
Because the DLL file must be compiled together with whatever program includes it.
Including an already compiled DLL file with optional arguments will cause crashes.
I have not tested it but this "feature" should exist in even .NET 7 as it is by design.

https://learn.microsoft.com/en-us/visualstudio/code-quality/ca1026
https://github.com/dotnet/roslyn-analyzers/issues/626
https://github.com/SonarSource/sonar-dotnet/issues/3622

Reply:
https://github.com/Metadorius/xna-cncnet-client/pull/10

Reply:
I don't understand how is this an issue though?

Reply:
As in we're not going to consume the library elsewhere, everything here always gets compiled in one sweep.

Reply:
> As in we're not going to consume the library elsewhere, everything here always gets compiled in one sweep.

If the two projects are coupled so tight, they should be merged into one whole project. `ClientCore` is a library.

Reply:
I absolutely disagree with this reasoning. Projects being split indicates logical separation of concerns. `ClientCore` implements core logic regardless of the UI that is used to display it, hence why it's a separate project.